src/service/layers-impl.js

```diff
@@ -0,0 +1,876 @@
+     * A later PR will refine this to send an array of elements who have
+     * changed position due to the scroll.
+     * @type {function()|null}
+     */
+    this.onScroll_ = null;
+
+    /** @const @private {!Array<!LayoutElement>} */
+    this.layouts_ = [];
+
+
+
+    // Listen for scroll events at the document level, so we know when either
+    // the document (the scrolling element) or an element scrolls.
+    // This forwards to our scroll-dirty system, and eventually to the scroll
+    // listener.
+    listen(this.document_, 'scroll', event => {
+      const {target} = event;
+      const scrolled = target.nodeType == Node.ELEMENT_NODE
+          ? target
+          : scrollingElement;
+      this.scrolled_(scrolled);
+    }, {capture: true, passive: true});
+
+    // Destroys the layer tree on document resize, since entirely new CSS may
+    // apply to the document now.
+    win.addEventListener('resize', () => this.onResize_());
+
+    // Finally, declare scrollingElement as the one true scrolling layer.
+    this.declareLayer_(scrollingElement, true);
+  }
+
+  /**
+   * Creates a layout for the element (if one doesn't exist for it already) and
+   * tracks the layout.
+   *
+   * This method may be called multiple times to ensure the element has a
+   * layout (and can therefore use the layout's instance methods) and that the
+   * layout will be remasured when necessary.
+   *
+   * @param {!Element} element
+   * @return {!Layout}
+   */
+  add(element) {
+    let layout = LayoutElement.forOptional(element);
+    // Elements may already have a layout (common for calls to get size or
+    // position from Resources).
+    if (!layout) {
+      layout = new LayoutElement(element);
+    }
+
+    // Layout may have been removed from the tracked layouts (due to
+    // reparenting).
+    if (this.layouts_.indexOf(layout) === -1) {
+      this.layouts_.push(layout);
+    }
+
+    return layout;
+  }
+
+  /**
+   * Removes the element's layout from tracking.
+   * This also "dirties" the layout, so if's being reparented it will lazily
+   * update appropriately.
+   *
+   * TODO(@jridgewell): This won't catch detach events from native DOM
+   * elements...
+   *
+   * @param {!Element} element
+   */
+  remove(element) {
+    const layout = LayoutElement.forOptional(element);
+    if (!layout) {
+      return;
+    }
+
+    const index = this.layouts_.indexOf(layout);
+    if (index > -1) {
+      this.layouts_.splice(index, 1);
+    }
+
+    const parent = layout.getParentLayer();
+    if (parent) {
+      parent.remove(layout);
+    }
+
+    // TODO child elements will disappear into the ether.
+    layout.undeclareLayer();
+  }
+
+  /**
+   * Returns the current scrolled position of the element relative to the layer
+   * represented by opt_parent (or opt_parent's parent layer, if it is not a
+   * layer itself). This takes into account the scrolled position of every
+   * layer in between.
```

Comment:
When is this called where `opt_parent` is not the direct parent? If that happens, we should call this `opt_ancestor`.

Reply:
Originally, it was for `<amp-carousel>` to determine the offset of one of its slides, which are actually contained in another layer. Renamed.